### ISSUE TITLE
Allow configuring optimizer_step based on the gradients.

### ISSUE THREAD

**Issue by rohitgr7** [PyTorchLightning/pytorch-lightning#5152]
As of now, there is no way to skip `optimizer_step` based on the gradients after they are calculated since the `backward` resides inside the closure and this closure is passed to the `optimizer.step`. Also if `training_step` returns `None` it still calls `optimizer_step` due to the closure which I think it should not if there are no pending accumulated gradients.

https://github.com/PyTorchLightning/pytorch-lightning/blob/fde972ffc2cdca3fccb904d29c2f1c32963fcd72/pytorch_lightning/trainer/training_loop.py#L707-L718
https://github.com/PyTorchLightning/pytorch-lightning/blob/fde972ffc2cdca3fccb904d29c2f1c32963fcd72/pytorch_lightning/core/optimizer.py#L127-L136

## Alternate option
- Use manual optimization
- zero_grad after backward the gradients and let optimizer_step happen which will have no effect.

## Expected behavior
Would like to have this support with automatic optimization with no hacky way (2nd point above)

**Comment by Borda** [comment]
@tchaton mind have a look :]

**Comment by tchaton** [comment]
Hey @rohitgr7,

After brainstorming with @carmocca, we feel there is no much we can do in `automatic_optimization` due to the design of Pytorch Optimizer.

```
    @torch.no_grad()
    def step(self, closure=None):
        """Performs a single optimization step.

        Arguments:
            closure (callable, optional): A closure that reevaluates the model
                and returns the loss.
        """
        loss = None
        if closure is not None:
            with torch.enable_grad():
                loss = closure()

        for group in self.param_groups:
```

**Comment by rohitgr7** [comment]
cool :+1: ... manual optimization to the rescue.
